File: frontend/src/types/insights.ts
/**
 * Quick Insights Types
 *
 * TypeScript types for automated data insights including
 * trend detection, outlier detection, and correlation analysis.
 */

// Enums

export type InsightType =
  | 'trend'
  | 'outlier'
  | 'correlation'
  | 'distribution'
  | 'comparison'
  | 'anomaly'
  | 'seasonality'
  | 'growth'
  | 'top_performer'
  | 'bottom_performer'
  | 'significant_change';

export type TrendDirection = 'increasing' | 'decreasing' | 'stable' | 'volatile';

export type InsightSeverity = 'high' | 'medium' | 'low';

export type CorrelationType = 'positive' | 'negative' | 'none';

export type OutlierType = 'above' | 'below' | 'both';

// Request Types

export interface InsightsRequest {
  connection_id: string;
  model_id?: string;
  table_name?: string;
  schema_name?: string;
  columns?: string[];
  date_column?: string;
  measure_columns?: string[];
  dimension_columns?: string[];
  limit_rows?: number;
  insight_types?: InsightType[];
}

export interface DatasetInsightsRequest {
  chart_id?: string;
  dataset_id?: string;
  data?: Record<string, unknown>[];
  columns?: string[];
  date_column?: string;
}

// Insight Detail Types

export interface TrendInsight {
  column: string;
  direction: TrendDirection;
  slope: number;
  r_squared: number;
  period: string;
  change_percent: number;
  start_value: number;
  end_value: number;
  data_points: number;
}

export interface OutlierInsight {
  column: string;
  outlier_type: OutlierType;
  values: Record<string, unknown>[];
  count: number;
  threshold_low?: number;
  threshold_high?: number;
  mean: number;
  std_dev: number;
  method: string;
}

export interface CorrelationInsight {
  column_1: string;
  column_2: string;
  correlation_type: CorrelationType;
  coefficient: number;
  p_value?: number;
  sample_size: number;
  interpretation: string;
}

export interface DistributionInsight {
  column: string;
  distribution_type: string;
  mean: number;
  median: number;
  mode?: number;
  std_dev: number;
  skewness: number;
  kurtosis: number;
  min_value: number;
  max_value: number;
  quartiles: number[];
}

export interface ComparisonInsight {
  measure: string;
  dimension: string;
  top_values: Record<string, unknown>[];
  bottom_values: Record<string, unknown>[];
  total: number;
  average: number;
  variance_coefficient: number;
}

export interface SeasonalityInsight {
  column: string;
  has_seasonality: boolean;
  period?: string;
  peak_periods: string[];
  trough_periods: string[];
  amplitude?: number;
}

export interface GrowthInsight {
  column: string;
  dimension_value?: string;
  period_1: string;
  period_2: string;
  value_1: number;
  value_2: number;
  absolute_change: number;
  percent_change: number;
  is_significant: boolean;
}

// Main Insight Type

export interface Insight {
  id: string;
  type: InsightType;
  severity: InsightSeverity;
  title: string;
  description: string;
  details: Record<string, unknown>;
  affected_columns: string[];
  confidence: number;
  generated_at: string;
  suggested_chart_type?: string;
  suggested_query?: string;
}

export interface InsightsResponse {
  insights: Insight[];
  summary: Record<string, unknown>;
  data_profile: Record<string, unknown>;
  execution_time_ms: number;
  rows_analyzed: number;
  columns_analyzed: number;
}

// Data Profile Types

export interface ColumnProfile {
  name: string;
  data_type: string;
  null_count: number;
  null_percent: number;
  unique_count: number;
  unique_percent: number;
  min_value?: number;
  max_value?: number;
  mean?: number;
  median?: number;
  std_dev?: number;
  min_length?: number;
  max_length?: number;
  avg_length?: number;
  top_values: Array<{
    value: string;
    count: number;
    percent: number;
  }>;
}

export interface DataProfile {
  row_count: number;
  column_count: number;
  columns: ColumnProfile[];
  memory_usage_bytes?: number;
  duplicate_rows: number;
  complete_rows: number;
}

// Trend Analysis Types

export interface TrendAnalysisRequest {
  connection_id: string;
  table_name: string;
  schema_name?: string;
  date_column: string;
  value_column: string;
  group_by?: string;
  granularity?: string;
  lookback_periods?: number;
}

export interface TrendAnalysisResponse {
  trend: TrendInsight;
  forecast: Record<string, unknown>[];
  seasonality?: SeasonalityInsight;
  data_points: Record<string, unknown>[];
}

// Outlier Detection Types

export interface OutlierDetectionRequest {
  connection_id: string;
  table_name: string;
  schema_name?: string;
  columns: string[];
  method?: 'iqr' | 'zscore' | 'isolation_forest';
  threshold?: number;
  zscore_threshold?: number;
}

export interface OutlierDetectionResponse {
  outliers: OutlierInsight[];
  total_outliers: number;
  outlier_percent: number;
}

// Correlation Analysis Types

export interface CorrelationAnalysisRequest {
  connection_id: string;
  table_name: string;
  schema_name?: string;
  columns: string[];
  method?: 'pearson' | 'spearman' | 'kendall';
  min_correlation?: number;
}

export interface CorrelationMatrix {
  columns: string[];
  matrix: number[][];
  significant_correlations: CorrelationInsight[];
}

// Constants

export const INSIGHT_TYPE_LABELS: Record<InsightType, string> = {
  trend: 'Trend',
  outlier: 'Outlier',
  correlation: 'Correlation',
  distribution: 'Distribution',
  comparison: 'Comparison',
  anomaly: 'Anomaly',
  seasonality: 'Seasonality',
  growth: 'Growth',
  top_performer: 'Top Performer',
  bottom_performer: 'Bottom Performer',
  significant_change: 'Significant Change',
};

export const INSIGHT_TYPE_ICONS: Record<InsightType, string> = {
  trend: 'trending-up',
  outlier: 'alert-triangle',
  correlation: 'link',
  distribution: 'bar-chart-2',
  comparison: 'git-compare',
  anomaly: 'zap',
  seasonality: 'calendar',
  growth: 'arrow-up-right',
  top_performer: 'award',
  bottom_performer: 'arrow-down',
  significant_change: 'activity',
};

export const SEVERITY_COLORS: Record<InsightSeverity, string> = {
  high: 'red',
  medium: 'yellow',
  low: 'blue',
};

export const TREND_DIRECTION_COLORS: Record<TrendDirection, string> = {
  increasing: 'green',
  decreasing: 'red',
  stable: 'gray',
  volatile: 'yellow',
};

export const CORRELATION_TYPE_COLORS: Record<CorrelationType, string> = {
  positive: 'green',
  negative: 'red',
  none: 'gray',
};

// Helper Functions

export function formatInsightSeverity(severity: InsightSeverity): string {
  return severity.charAt(0).toUpperCase() + severity.slice(1);
}

export function getInsightIcon(type: InsightType): string {
  return INSIGHT_TYPE_ICONS[type] || 'info';
}

export function getSeverityColor(severity: InsightSeverity): string {
  return SEVERITY_COLORS[severity];
}

export function formatConfidence(confidence: number): string {
  return `${Math.round(confidence * 100)}%`;
}

export function formatPercentChange(change: number): string {
  const sign = change >= 0 ? '+' : '';
  return `${sign}${change.toFixed(1)}%`;
}

export function sortInsightsBySeverity(insights: Insight[]): Insight[] {
  const severityOrder: Record<InsightSeverity, number> = {
    high: 0,
    medium: 1,
    low: 2,
  };

  return [...insights].sort((a, b) => {
    const severityDiff = severityOrder[a.severity] - severityOrder[b.severity];
    if (severityDiff !== 0) return severityDiff;
    return b.confidence - a.confidence;
  });
}

export function filterInsightsByType(
  insights: Insight[],
  types: InsightType[]
): Insight[] {
  return insights.filter((insight) => types.includes(insight.type));
}

export function groupInsightsByType(
  insights: Insight[]
): Record<InsightType, Insight[]> {
  const grouped: Record<string, Insight[]> = {};

  for (const insight of insights) {
    if (!grouped[insight.type]) {
      grouped[insight.type] = [];
    }
    grouped[insight.type].push(insight);
  }

  return grouped as Record<InsightType, Insight[]>;
}

export function getHighPriorityInsights(insights: Insight[]): Insight[] {
  return insights.filter((insight) => insight.severity === 'high');
}

export function createAnalysisRequest(
  data: Record<string, unknown>[],
  options?: {
    columns?: string[];
    dateColumn?: string;
    measureColumns?: string[];
    dimensionColumns?: string[];
    insightTypes?: InsightType[];
  }
): Record<string, unknown> {
  return {
    data,
    columns: options?.columns,
    date_column: options?.dateColumn,
    measure_columns: options?.measureColumns,
    dimension_columns: options?.dimensionColumns,
    insight_types: options?.insightTypes,
  };
}

File: backend/app/schemas/insights.py
"""
Quick Insights Schemas

Defines data models for automated data insights including
trend detection, outlier detection, correlation analysis, and more.
"""

from typing import Any, Optional
from pydantic import BaseModel, Field
from enum import Enum
from datetime import datetime


# Enums

class InsightType(str, Enum):
    """Type of insight detected"""
    TREND = "trend"
    OUTLIER = "outlier"
    CORRELATION = "correlation"
    DISTRIBUTION = "distribution"
    COMPARISON = "comparison"
    ANOMALY = "anomaly"
    SEASONALITY = "seasonality"
    GROWTH = "growth"
    TOP_PERFORMER = "top_performer"
    BOTTOM_PERFORMER = "bottom_performer"
    SIGNIFICANT_CHANGE = "significant_change"


class TrendDirection(str, Enum):
    """Direction of a trend"""
    INCREASING = "increasing"
    DECREASING = "decreasing"
    STABLE = "stable"
    VOLATILE = "volatile"


class InsightSeverity(str, Enum):
    """Severity/importance of an insight"""
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


class CorrelationType(str, Enum):
    """Type of correlation"""
    POSITIVE = "positive"
    NEGATIVE = "negative"
    NONE = "none"


class OutlierType(str, Enum):
    """Type of outlier"""
    ABOVE = "above"
    BELOW = "below"
    BOTH = "both"


# Request Models

class InsightsRequest(BaseModel):
    """Request for insights analysis"""
    connection_id: str
    model_id: Optional[str] = None
    table_name: Optional[str] = None
    schema_name: str = "public"
    columns: list[str] = Field(default_factory=list)  # Specific columns to analyze
    date_column: Optional[str] = None  # For time-based analysis
    measure_columns: list[str] = Field(default_factory=list)  # Numeric columns for analysis
    dimension_columns: list[str] = Field(default_factory=list)  # Categorical columns
    limit_rows: int = Field(default=10000, le=100000)
    insight_types: list[InsightType] = Field(default_factory=list)  # Filter specific types


class DatasetInsightsRequest(BaseModel):
    """Request for insights on a specific dataset/chart"""
    chart_id: Optional[str] = None
    dataset_id: Optional[str] = None
    data: Optional[list[dict[str, Any]]] = None  # Pre-loaded data
    columns: list[str] = Field(default_factory=list)
    date_column: Optional[str] = None


# Insight Models

class TrendInsight(BaseModel):
    """A trend insight"""
    column: str
    direction: TrendDirection
    slope: float
    r_squared: float  # How well the trend fits
    period: str  # e.g., "last 30 days", "Q1 2024"
    change_percent: float
    start_value: float
    end_value: float
    data_points: int


class OutlierInsight(BaseModel):
    """An outlier insight"""
    column: str
    outlier_type: OutlierType
    values: list[dict[str, Any]]  # The outlier records
    count: int
    threshold_low: Optional[float] = None
    threshold_high: Optional[float] = None
    mean: float
    std_dev: float
    method: str = "iqr"  # iqr, zscore, isolation_forest


class CorrelationInsight(BaseModel):
    """A correlation insight"""
    column_1: str
    column_2: str
    correlation_type: CorrelationType
    coefficient: float  # -1 to 1
    p_value: Optional[float] = None
    sample_size: int
    interpretation: str


class DistributionInsight(BaseModel):
    """A distribution insight"""
    column: str
    distribution_type: str  # normal, skewed_left, skewed_right, uniform, bimodal
    mean: float
    median: float
    mode: Optional[float] = None
    std_dev: float
    skewness: float
    kurtosis: float
    min_value: float
    max_value: float
    quartiles: list[float]  # Q1, Q2, Q3


class ComparisonInsight(BaseModel):
    """A comparison insight"""
    measure: str
    dimension: str
    top_values: list[dict[str, Any]]
    bottom_values: list[dict[str, Any]]
    total: float
    average: float
    variance_coefficient: float


class SeasonalityInsight(BaseModel):
    """A seasonality insight"""
    column: str
    has_seasonality: bool
    period: Optional[str] = None  # daily, weekly, monthly, quarterly, yearly
    peak_periods: list[str] = Field(default_factory=list)
    trough_periods: list[str] = Field(default_factory=list)
    amplitude: Optional[float] = None


class GrowthInsight(BaseModel):
    """A growth/change insight"""
    column: str
    dimension_value: Optional[str] = None
    period_1: str
    period_2: str
    value_1: float
    value_2: float
    absolute_change: float
    percent_change: float
    is_significant: bool


class Insight(BaseModel):
    """A single insight"""
    id: str
    type: InsightType
    severity: InsightSeverity
    title: str
    description: str
    details: dict[str, Any] = Field(default_factory=dict)  # Type-specific details
    affected_columns: list[str] = Field(default_factory=list)
    confidence: float = Field(default=0.8, ge=0, le=1)
    generated_at: datetime = Field(default_factory=datetime.utcnow)

    # Visualization suggestion
    suggested_chart_type: Optional[str] = None
    suggested_query: Optional[str] = None


class InsightsResponse(BaseModel):
    """Response with all insights"""
    insights: list[Insight]
    summary: dict[str, Any] = Field(default_factory=dict)
    data_profile: dict[str, Any] = Field(default_factory=dict)
    execution_time_ms: float
    rows_analyzed: int
    columns_analyzed: int


# Data Profile Models

class ColumnProfile(BaseModel):
    """Profile of a single column"""
    name: str
    data_type: str
    null_count: int
    null_percent: float
    unique_count: int
    unique_percent: float

    # For numeric columns
    min_value: Optional[float] = None
    max_value: Optional[float] = None
    mean: Optional[float] = None
    median: Optional[float] = None
    std_dev: Optional[float] = None

    # For string columns
    min_length: Optional[int] = None
    max_length: Optional[int] = None
    avg_length: Optional[float] = None

    # Top values
    top_values: list[dict[str, Any]] = Field(default_factory=list)


class DataProfile(BaseModel):
    """Complete data profile"""
    row_count: int
    column_count: int
    columns: list[ColumnProfile]
    memory_usage_bytes: Optional[int] = None
    duplicate_rows: int = 0
    complete_rows: int = 0  # Rows with no nulls


# Trend Analysis Models

class TrendAnalysisRequest(BaseModel):
    """Request for trend analysis"""
    connection_id: str
    table_name: str
    schema_name: str = "public"
    date_column: str
    value_column: str
    group_by: Optional[str] = None
    granularity: str = "day"  # day, week, month, quarter, year
    lookback_periods: int = 30


class TrendAnalysisResponse(BaseModel):
    """Response with trend analysis"""
    trend: TrendInsight
    forecast: list[dict[str, Any]] = Field(default_factory=list)
    seasonality: Optional[SeasonalityInsight] = None
    data_points: list[dict[str, Any]]


# Outlier Detection Models

class OutlierDetectionRequest(BaseModel):
    """Request for outlier detection"""
    connection_id: str
    table_name: str
    schema_name: str = "public"
    columns: list[str]
    method: str = "iqr"  # iqr, zscore, isolation_forest
    threshold: float = 1.5  # For IQR method
    zscore_threshold: float = 3.0  # For z-score method


class OutlierDetectionResponse(BaseModel):
    """Response with outlier detection"""
    outliers: list[OutlierInsight]
    total_outliers: int
    outlier_percent: float


# Correlation Analysis Models

class CorrelationAnalysisRequest(BaseModel):
    """Request for correlation analysis"""
    connection_id: str
    table_name: str
    schema_name: str = "public"
    columns: list[str]  # At least 2 columns
    method: str = "pearson"  # pearson, spearman, kendall
    min_correlation: float = 0.5  # Minimum correlation to report


class CorrelationMatrix(BaseModel):
    """Correlation matrix result"""
    columns: list[str]
    matrix: list[list[float]]
    significant_correlations: list[CorrelationInsight]


# Constants

INSIGHT_TEMPLATES = {
    InsightType.TREND: {
        "increasing": "{column} shows an increasing trend of {change_percent:.1f}% over {period}",
        "decreasing": "{column} shows a decreasing trend of {change_percent:.1f}% over {period}",
        "stable": "{column} has remained stable over {period} with minimal variation",
        "volatile": "{column} shows high volatility over {period}",
    },
    InsightType.OUTLIER: {
        "above": "Found {count} values in {column} significantly above the normal range",
        "below": "Found {count} values in {column} significantly below the normal range",
        "both": "Found {count} outliers in {column} both above and below normal range",
    },
    InsightType.CORRELATION: {
        "positive": "{column_1} and {column_2} are strongly positively correlated (r={coefficient:.2f})",
        "negative": "{column_1} and {column_2} are strongly negatively correlated (r={coefficient:.2f})",
    },
    InsightType.TOP_PERFORMER: {
        "default": "Top performing {dimension}: {top_value} with {measure}={value:.2f}",
    },
    InsightType.SIGNIFICANT_CHANGE: {
        "increase": "{column} increased by {change_percent:.1f}% from {period_1} to {period_2}",
        "decrease": "{column} decreased by {change_percent:.1f}% from {period_1} to {period_2}",
    },
}

CHART_SUGGESTIONS = {
    InsightType.TREND: "line",
    InsightType.OUTLIER: "scatter",
    InsightType.CORRELATION: "scatter",
    InsightType.DISTRIBUTION: "histogram",
    InsightType.COMPARISON: "bar",
    InsightType.TOP_PERFORMER: "bar",
    InsightType.SEASONALITY: "line",
    InsightType.GROWTH: "bar",
}

SEVERITY_THRESHOLDS = {
    "trend_change_percent": {"high": 25, "medium": 10},
    "outlier_percent": {"high": 5, "medium": 2},
    "correlation_coefficient": {"high": 0.8, "medium": 0.6},
    "growth_percent": {"high": 50, "medium": 20},
}
